File: client/lib/analytics/play_simulation.dart
/// Kliensoldali figura-szimuláció — a backend play_simulation.py Dart-tükre.
///
/// Megtanulja egy csapat védekezési stílusát, lejátssza az edző figuráját a
/// tanult védelem ellen (Tracking kimenet), és pontozza a teremtett lövőhelyzetet.
/// Kliensoldalon, így backend nélkül (a betöltött/demó meccsből tanulva) is megy.
library;

import "dart:math" as math;
import "dart:ui";

import "../models/tracking.dart";
import "../ui/court_geometry.dart";
import "decisions.dart" show shotValue;
import "tactics.dart";

/// Egy csapat tanult védekezési stílusa.
class DefenseModel {
  final int numDefenders;
  final double lineDepthM;  // a védővonal mélysége a saját kaputól
  final double lateralGain; // mennyire követi a labda y-helyzetét

  const DefenseModel({this.numDefenders = 6, this.lineDepthM = 6.0, this.lateralGain = 0.5});

  /// A védők pozíciói a labdára reagálva (a tanult mélységben, y-ban a labda felé).
  List<Offset> respond(double ballY, double goalX) {
    final sign = goalX == courtLength ? -1.0 : 1.0;
    final lineX = goalX + sign * lineDepthM;
    final n = math.max(1, numDefenders);
    final baseYs = <double>[];
    if (n == 1) {
      baseYs.add(courtWidth / 2);
    } else {
      const lo = 3.0;
      final hi = courtWidth - 3.0;
      for (int i = 0; i < n; i++) {
        baseYs.add(lo + (hi - lo) * i / (n - 1));
      }
    }
    final shift = lateralGain * (ballY - courtWidth / 2);
    return [for (final by in baseYs) Offset(lineX, (by + shift).clamp(1.0, courtWidth - 1.0).toDouble())];
  }
}

/// Védekezési stílus tanulása egy meccsből (a defending csapat védőfázisaiból).
DefenseModel learnDefense(Match match, Team team,
    {TacticsConfig config = const TacticsConfig()}) {
  final goalX = config.ownGoalX(team);
  final depths = <double>[];
  final counts = <int>[];
  for (final f in match.frames) {
    final ph = classifyPhase(f, config);
    final defends = (team == Team.away && ph == Phase.homeAttack) ||
        (team == Team.home && ph == Phase.awayAttack);
    if (!defends) continue;
    final outfield = f.players.where((p) => p.team == team && (p.x - goalX).abs() > 2.0).toList();
    if (outfield.isNotEmpty) {
      counts.add(outfield.length);
      depths.addAll(outfield.map((p) => (p.x - goalX).abs()));
    }
  }
  double mean(List<double> v) => v.isEmpty ? 0 : v.reduce((a, b) => a + b) / v.length;
  return DefenseModel(
    numDefenders: counts.isEmpty ? 6 : (counts.reduce((a, b) => a + b) / counts.length).round(),
    lineDepthM: depths.isEmpty ? 6.0 : mean(depths),
  );
}

/// Az edző figurája: támadók útvonala lépésenként + ki birtokolja a labdát.
class SetPlay {
  final List<List<Offset>> attackers; // [támadó][lépés] = (x,y) méter
  final List<int> ballCarrier;        // lépésenként a labdás támadó indexe
  const SetPlay(this.attackers, this.ballCarrier);
  int get steps => ballCarrier.length;
}

/// Lejátssza a figurát a tanult védelem ellen → Match (Tracking).
Match simulateSetPlay(SetPlay setplay, DefenseModel defense,
    {TacticsConfig config = const TacticsConfig(), double fps = 25.0}) {
  final awayGoalX = config.ownGoalX(Team.away);
  final meta = MatchMeta(
    matchId: "setplay-sim", homeTeam: "Terv (támadó)", awayTeam: "Tanult védelem",
    fps: fps, frameWidth: 1920, frameHeight: 1080,
  );
  final frames = <Frame>[];
  for (int step = 0; step < setplay.steps; step++) {
    final players = <PlayerPosition>[];
    for (int ai = 0; ai < setplay.attackers.length; ai++) {
      final pt = setplay.attackers[ai][step];
      players.add(PlayerPosition(
        trackId: ai + 1, team: Team.home, x: pt.dx, y: pt.dy,
        source: PositionSource.measured, confidence: 1.0, jerseyNumber: ai + 1,
      ));
    }
    final carrier = setplay.ballCarrier[step];
    final bp = setplay.attackers[carrier][step];
    final ball = Ball(x: bp.dx, y: bp.dy, confidence: 1.0);
    final defs = defense.respond(bp.dy, awayGoalX);
    for (int di = 0; di < defs.length; di++) {
      players.add(PlayerPosition(
        trackId: 100 + di, team: Team.away, x: defs[di].dx, y: defs[di].dy,
        source: PositionSource.measured, confidence: 1.0,
      ));
    }
    frames.add(Frame(t: step, players: players, ball: ball));
  }
  return Match(meta: meta, frames: frames);
}

/// A szimulált figura kiértékelése: a teremtett legjobb lövőhelyzet.
class SetPlayEvaluation {
  final double bestShotValue;
  final int step;
  final int? attackerId;
  const SetPlayEvaluation(this.bestShotValue, this.step, this.attackerId);
}

SetPlayEvaluation evaluateSetPlay(Match match, {TacticsConfig config = const TacticsConfig()}) {
  final goalX = config.attacksTowardX(Team.home);
  double best = 0.0;
  int bestStep = -1;
  int? bestAtk;
  for (final f in match.frames) {
    final attackers = f.players.where((p) => p.team == Team.home);
    final defenders = f.players.where((p) => p.team == Team.away).toList();
    for (final a in attackers) {
      final sv = shotValue(a.x, a.y, goalX);
      double nd = 99.0;
      for (final d in defenders) {
        final dd = math.sqrt(math.pow(a.x - d.x, 2) + math.pow(a.y - d.y, 2));
        if (dd < nd) nd = dd;
      }
      final openness = (nd / 4.0).clamp(0.0, 1.0);
      final score = sv * openness;
      if (score > best) {
        best = score;
        bestStep = f.t;
        bestAtk = a.trackId;
      }
    }
  }
  return SetPlayEvaluation(best, bestStep, bestAtk);
}

File: backend/handball/pipeline/play_simulation.py
"""
[5. fázis] Figura-szimuláció — az edző figurája egy TANULT ellenfél ellen.

A vízió "edző kipróbálja a kitalált figuráit egy adott csapat ellen" része:
1. DefenseModel.learn(...) — egy ellenfél meccséből megtanuljuk a védekezési
   stílusát (hány védő, milyen mélyen áll a saját kaputól, mennyire követi a labdát).
2. SetPlay — az edző által megtervezett figura: a támadók útvonala lépésenként +
   ki birtokolja a labdát.
3. simulate_setplay(...) — lejátssza a figurát: a támadók a terv szerint mozognak,
   a védők a TANULT modell szerint reagálnak. Kimenet egy Match (Tracking), amit a
   kliens ugyanúgy ki tud rajzolni, mint egy valódi meccset.
4. evaluate_setplay(...) — pontozza a figurát: teremtett-e szabad, közeli
   lövőhelyzetet (a legjobb lövésérték * "szabadság" a védőktől).

Tiszta Python, videó nélkül tesztelhető. A védekezési modell egyszerű, de VALÓDI
adatból tanult paraméterekre épül — később finomabb (tanult) modellre cserélhető.
"""

from __future__ import annotations

import math
import statistics
from dataclasses import dataclass, field

from ..models.tracking import (
    Match, MatchMeta, Frame, PlayerPosition, Ball, Team, PositionSource,
)
from .calibration import COURT_LENGTH_M, COURT_WIDTH_M
from .tactics import TacticsConfig, classify_phase, Phase
from .decisions import shot_value


# ---- Tanult védekezési modell ---------------------------------------------

@dataclass
class DefenseModel:
    """Egy csapat védekezési stílusa, meccsadatból tanult paraméterekkel.

    - num_defenders: hány mezőnyvédő áll a vonalban (kapus nélkül).
    - line_depth_m:  a védővonal átlagos mélysége a SAJÁT kaputól (méter).
    - lateral_gain:  mennyire követi a védelem a labda y-helyzetét (0..1).
    """
    num_defenders: int = 6
    line_depth_m: float = 6.0
    lateral_gain: float = 0.5

    @classmethod
    def learn(cls, match: Match, defending_team: Team,
              config: TacticsConfig | None = None) -> "DefenseModel":
        """Megtanulja a védekezési stílust azokból a frame-ekből, ahol a csapat véd.

        A csapat akkor véd, amikor az ellenfél támad. Ezeken a frame-eken a
        mezőnyvédők (kapus nélkül) átlagos mélységét és átlagos számát becsüljük.
        """
        config = config or TacticsConfig()
        goal_x = config.own_goal_x(defending_team)
        depths: list[float] = []
        counts: list[int] = []
        for f in match.frames:
            ph = classify_phase(f, config)
            defends = ((defending_team == Team.AWAY and ph == Phase.HOME_ATTACK) or
                       (defending_team == Team.HOME and ph == Phase.AWAY_ATTACK))
            if not defends:
                continue
            outfield = [p for p in f.players
                        if p.team == defending_team and abs(p.x - goal_x) > 2.0]
            if outfield:
                counts.append(len(outfield))
                depths.extend(abs(p.x - goal_x) for p in outfield)
        return cls(
            num_defenders=round(statistics.fmean(counts)) if counts else 6,
            line_depth_m=statistics.fmean(depths) if depths else 6.0,
        )

    def respond(self, ball: Ball, goal_x: float) -> list[tuple[float, float]]:
        """A védők pozíciói a labdára reagálva (a tanult mélységben, y-ban a labda felé).

        A védővonal a kaputól `line_depth_m`-re húzódik (a pálya belseje felé), a
        védők y-ban egyenletesen elosztva, a labda y-helyzete felé eltolva.
        """
        sign = -1.0 if goal_x == COURT_LENGTH_M else 1.0
        line_x = goal_x + sign * self.line_depth_m

        n = max(1, self.num_defenders)
        # Egyenletes y-eloszlás a [3, 17] sávban.
        if n == 1:
            base_ys = [COURT_WIDTH_M / 2.0]
        else:
            lo, hi = 3.0, COURT_WIDTH_M - 3.0
            base_ys = [lo + (hi - lo) * i / (n - 1) for i in range(n)]

        shift = self.lateral_gain * (ball.y - COURT_WIDTH_M / 2.0)
        return [(line_x, max(1.0, min(COURT_WIDTH_M - 1.0, by + shift))) for by in base_ys]


# ---- A figura (set play) és a szimuláció ----------------------------------

@dataclass
class SetPlay:
    """Az edző által tervezett figura.

    - attackers:    [támadó][lépés] = (x, y) — minden támadó útvonala lépésenként.
                    Minden támadó listája azonos hosszú (a lépések száma).
    - ball_carrier: lépésenként melyik támadó (index) birtokolja a labdát.
    """
    attackers: list[list[tuple[float, float]]]
    ball_carrier: list[int]

    @property
    def steps(self) -> int:
        return len(self.ball_carrier)


def simulate_setplay(setplay: SetPlay, defense: DefenseModel,
                     config: TacticsConfig | None = None,
                     fps: float = 25.0) -> Match:
    """Lejátssza a figurát a tanult védelem ellen, és Match-et (Tracking) ad.

    A HAZAI a támadó (a +x kapu felé), a VENDÉG véd (saját kapuja x=40). A
    kimenet ugyanolyan Tracking, mint egy valódi meccs — a kliens kirajzolhatja.
    """
    config = config or TacticsConfig()
    away_goal_x = config.own_goal_x(Team.AWAY)
    meta = MatchMeta(match_id="setplay-sim", home_team="Terv (támadó)",
                     away_team="Tanult védelem", fps=fps, frame_width=1920, frame_height=1080)
    frames: list[Frame] = []

    for step in range(setplay.steps):
        players: list[PlayerPosition] = []
        for ai, path in enumerate(setplay.attackers):
            x, y = path[step]
            players.append(PlayerPosition(track_id=ai + 1, team=Team.HOME, x=x, y=y,
                                          source=PositionSource.MEASURED, confidence=1.0,
                                          jersey_number=ai + 1))
        carrier = setplay.ball_carrier[step]
        bx, by = setplay.attackers[carrier][step]
        ball = Ball(x=bx, y=by, confidence=1.0)
        for di, (dx, dy) in enumerate(defense.respond(ball, away_goal_x)):
            players.append(PlayerPosition(track_id=100 + di, team=Team.AWAY, x=dx, y=dy,
                                          source=PositionSource.MEASURED, confidence=1.0))
        frames.append(Frame(t=step, players=players, ball=ball))

    return Match(meta=meta, frames=frames)


def evaluate_setplay(match: Match, config: TacticsConfig | None = None) -> dict:
    """Pontozza a szimulált figurát: a teremtett legjobb lövőhelyzet.

    Minden lépésben minden támadóra: lövésérték * "szabadság" (a legközelebbi védő
    távolsága alapján — 4 m-en belül fedezett). A legjobb érték a figura pontszáma.
    """
    config = config or TacticsConfig()
    goal_x = config.attacks_toward_x(Team.HOME)
    best = {"best_shot_value": 0.0, "step": -1, "attacker_id": None}

    for f in match.frames:
        attackers = [p for p in f.players if p.team == Team.HOME]
        defenders = [p for p in f.players if p.team == Team.AWAY]
        for a in attackers:
            sv = shot_value(a.x, a.y, goal_x)
            if defenders:
                nd = min(math.hypot(a.x - d.x, a.y - d.y) for d in defenders)
            else:
                nd = 99.0
            openness = max(0.0, min(1.0, nd / 4.0))  # 4 m-en belül fedezett
            score = sv * openness
            if score > best["best_shot_value"]:
                best = {"best_shot_value": score, "step": f.t, "attacker_id": a.track_id}
    return best
